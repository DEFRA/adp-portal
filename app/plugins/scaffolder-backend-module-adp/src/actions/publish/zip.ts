import {
  getRepoSourceDirectory,
  createTemplateAction,
} from '@backstage/plugin-scaffolder-node';
import archiver, { type Archiver } from 'archiver';
import type { Readable } from 'node:stream';
import type { Logger } from 'winston';

export type PublishZipActionInput = {
  readonly sourcePath?: string;
};
export type PublishZipActionOutput = {
  readonly dataUri: string;
};
export const publishZipAction = createTemplateAction<
  PublishZipActionInput,
  PublishZipActionOutput
>({
  id: 'publish:zip',
  description:
    'Publishes the scaffolded repo to a zip file which can be downloaded.',
  schema: {
    input: {
      type: 'object',
      properties: {
        sourcePath: {
          title: 'Source Path',
          description:
            'Path within the workspace that will be used as the repository root. If omitted, the entire workspace will be published as the repository.',
          type: 'string',
        },
      },
    },
    output: {
      type: 'object',
      properties: {
        dataUri: {
          title: 'A data uri which contains the zipped contents',
          type: 'string',
        },
      },
    },
  },
  async handler(ctx) {
    const dir = getRepoSourceDirectory(ctx.workspacePath, ctx.input.sourcePath);

    const data = await createArchive(
      opt =>
        opt.directory(dir, false, {
          date: new Date(),
        }),
      {
        logger: ctx.logger,
      },
    );
    ctx.output(
      'dataUri',
      `data:application/zip;base64,${data.toString('base64')}`,
    );
  },
});

interface CreateArchiveOptions extends archiver.ArchiverOptions {
  readonly format?: archiver.Format;
  readonly logger?: Logger;
}

async function createArchive(
  populate: (archive: Archiver) => void,
  options: CreateArchiveOptions = {},
) {
  const archive = archiver(options.format ?? 'zip', options);
  using _onWarning = usingListener(archive, 'warning', (err: unknown) =>
    options.logger?.warn(err),
  );
  using _onError = usingListener(archive, 'error', (err: unknown) =>
    options.logger?.error(err),
  );
  const result = getData(archive);
  populate(archive);
  await archive.finalize();
  return await result;
}

async function getData(stream: Readable) {
  const chunks: Uint8Array[] = [];
  for await (const chunk of stream) {
    chunks.push(chunk);
  }
  return Buffer.concat(chunks);
}

function usingListener<TEvent, THandler>(
  target: { [P in 'on' | 'off']: (event: TEvent, handler: THandler) => any },
  event: TEvent,
  handler: THandler,
) {
  target.on(event, handler);
  let disposed = false;
  return {
    [Symbol.dispose]() {
      if (disposed) return;
      disposed = true;
      target.off(event, handler);
    },
  };
}

if (!Symbol.dispose) {
  Object.defineProperty(Symbol, 'dispose', {
    value: Symbol('Symbol.dispose'),
    writable: false,
    configurable: false,
  });
}
