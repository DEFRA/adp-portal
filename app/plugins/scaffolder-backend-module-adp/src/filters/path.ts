import type { JsonValue } from '@backstage/types';
import pathImpl from 'node:path';

export function path(input: JsonValue, ...segments: JsonValue[]) {
  const baseStr = String(input);
  const schemeMatch = /^\w+:\//.exec(baseStr);
  const [scheme, s1] =
    schemeMatch === null
      ? ['', baseStr]
      : [
          baseStr.slice(0, schemeMatch[0].length),
          baseStr.slice(schemeMatch[0].length),
        ];

  return scheme + pathImpl.join(s1, ...segments.map(String));
}
