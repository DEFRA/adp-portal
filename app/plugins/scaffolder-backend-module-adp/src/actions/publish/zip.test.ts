import type { ActionContext } from '@backstage/plugin-scaffolder-node';
import { publishZipAction } from './zip';
import { getVoidLogger } from '@backstage/backend-common';
import { Writable } from 'node:stream';
import path from 'node:path';

jest.useFakeTimers({
  now: new Date('2024-06-06T21:13:57.075Z'),
  doNotFake: [
    'clearImmediate',
    'clearTimeout',
    'nextTick',
    'setImmediate',
    'setTimeout',
  ],
});

describe('publishZipAction', () => {
  it('Should zip the source directory', async () => {
    // arrange
    const context: jest.Mocked<ActionContext<{ sourcePath?: string }>> = {
      createTemporaryDirectory: jest.fn(),
      input: {
        sourcePath: undefined,
      },
      logger: getVoidLogger(),
      logStream: new Writable(),
      output: jest.fn(),
      workspacePath: path.join(__dirname, '__testdata__'),
    };

    // act
    await publishZipAction.handler(context);

    // assert
    expect(context.output).toHaveBeenCalledTimes(1);
    expect(context.output.mock.calls[0]).toMatchSnapshot();
  });
});
