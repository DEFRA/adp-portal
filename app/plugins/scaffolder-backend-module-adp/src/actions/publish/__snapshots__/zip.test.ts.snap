// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`publishZipAction Should zip the source directory 1`] = `
[
  "dataUri",
  "data:application/zip;base64,UEsDBBQACAAIALypxlgAAAAAAAAAAAAAAAAIAAAAdGVzdC50eHQFwcENgDAMA8A/U5idWCBqHRKpqFVjmJ+7K7KQBYNYgucgFCYoiGrmPkfnxpN3CGEfsfbsb2M/jx9QSwcI0eA/qjkAAAA9AAAAUEsBAi0DFAAIAAgAvKnGWNHgP6o5AAAAPQAAAAgAAAAAAAAAAAAgAKSBAAAAAHRlc3QudHh0UEsFBgAAAAABAAEANgAAAG8AAAAAAA==",
]
`;
